lib/auth/keystore/gcp_kms_test.go

```diff
@@ -0,0 +1,367 @@
+	go func() {
+		defer m.wg.Done()
+
+		<-m.activateKey
+		defer func() { m.keyActivated <- struct{}{} }()
+
+		m.Lock()
+		defer m.Unlock()
+		m.keyVersions[req.CryptoKeyId+keyVersionSuffix] = "ACTIVE"
+	}()
+
+	return &kmspb.CryptoKey{
+		Name: req.CryptoKeyId,
+	}, nil
+}
+
+func (m *mockGCPKMSClient) DestroyCryptoKeyVersion(ctx context.Context, req *kmspb.DestroyCryptoKeyVersionRequest, opts ...gax.CallOption) (*kmspb.CryptoKeyVersion, error) {
+	// Don't return until someone sends to unblockDestroyCryptoKey or the
+	// context expires.
+	select {
+	case <-m.unblockDestroyCryptoKey:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+	m.Lock()
+	defer m.Unlock()
+	if m.keyVersions[req.Name] != "ACTIVE" {
+		return nil, errors.New("DestroyCryptoKeyVersion failed, state has value " + m.keyVersions[req.Name])
+	}
+	m.keyVersions[req.Name] = "DESTROY_SCHEDULED"
+	m.keyIsDestroyed <- struct{}{}
+	return nil, nil
+}
+
+func (m *mockGCPKMSClient) ListCryptoKeys(ctx context.Context, req *kmspb.ListCryptoKeysRequest, opts ...gax.CallOption) cryptoKeyIteratorI {
+	m.RLock()
+	defer m.RUnlock()
+	var keys []string
+	for keyVersion := range m.keyVersions {
+		keyName := keyVersion[:len(keyVersion)-len(keyVersionSuffix)]
+		keys = append(keys, keyName)
+	}
+	return &mockCryptoKeyIterator{
+		keys:            keys,
+		ctx:             ctx,
+		unblockIterator: m.unblockIterator,
+	}
+}
+
+func (m *mockGCPKMSClient) GetPublicKey(ctx context.Context, req *kmspb.GetPublicKeyRequest, opts ...gax.CallOption) (*kmspb.PublicKey, error) {
+	// Don't return until someone sends to unblockGetPublicKey or the context
+	// expires. Defer this so that we the state can be determistically changed
+	// between calls. If it were not deferred, if a caller were to unblock this
+	// and then change the state, it could possibly be reflected in this call or
+	// the next.
+	defer func() {
+		select {
+		case <-m.unblockGetPublicKey:
+		case <-ctx.Done():
+		}
+	}()
+	m.RLock()
+	defer m.RUnlock()
+	if m.keyVersions[req.Name] != "ACTIVE" {
+		return nil, errors.New("GetPublicKey failed, state has value " + m.keyVersions[req.Name])
+	}
+	return &kmspb.PublicKey{
+		Pem: `-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAryQICCl6NZ5gDKrnSztO
+3Hy8PEUcuyvg/ikC+VcIo2SFFSf18a3IMYldIugqqqZCs4/4uVW3sbdLs/6PfgdX
+7O9D22ZiFWHPYA2k2N744MNiCD1UE+tJyllUhSblK48bn+v1oZHCM0nYQ2NqUkvS
+j+hwUU3RiWl7x3D2s9wSdNt7XUtW05a/FXehsPSiJfKvHJJnGOX0BgTvkLnkAOTd
+OrUZ/wK69Dzu4IvrN4vs9Nes8vbwPa/ddZEzGR0cQMt0JBkhk9kU/qwqUseP1QRJ
+5I1jR4g8aYPL/ke9K35PxZWuDp3U0UPAZ3PjFAh+5T+fc7gzCs9dPzSHloruU+gl
+FQIDAQAB
+-----END PUBLIC KEY-----`,
+	}, nil
+}
+
+// Never actually succeeds, just blocks until the context is cancelled
+func (m *mockGCPKMSClient) AsymmetricSign(ctx context.Context, req *kmspb.AsymmetricSignRequest, opts ...gax.CallOption) (*kmspb.AsymmetricSignResponse, error) {
+	for {
+		select {
+		case m.signIsBlocked <- struct{}{}:
+		case <-ctx.Done():
+			return nil, ctx.Err()
```

Comment:
This looks... off? It never returns unless the context expires, and always errors?

Reply:
`ctx.Err()` can be nil, I think

Reply:
Not if the context is closed, which is guaranteed by the `<-ctx.Done()` above (see Done() and Err() - https://pkg.go.dev/context#Context).